yt_idefix/data_structures.py

```diff
@@ -427,15 +427,45 @@ def _get_time(self):
 
     def _set_code_unit_attributes(self):
         """Conversion between physical units and code units."""
+        # Get units defined in definitions.h
         length_unit = self.parameters.get("unit_length", 1.0)
         velocity_unit = self.parameters.get("unit_velocity", 1.0)
         density_unit = self.parameters.get("unit_density", 1.0)
 
-        self.length_unit = self.quan(length_unit, "cm")
-        self.velocity_unit = self.quan(velocity_unit, "cm/s")
-        density_unit = self.quan(density_unit, "g/cm**3")
+        # Get override_units
+        uo = self.units_override
+        self.length_unit = uo.get("length_unit", self.quan(length_unit, "cm"))
+        self.velocity_unit = uo.get("velocity_unit", self.quan(velocity_unit, "cm/s"))
+        density_unit = uo.get("density_unit", self.quan(density_unit, "g/cm**3"))
 
         self.mass_unit = density_unit * self.length_unit ** 3
         self.time_unit = self.length_unit / self.velocity_unit
         self.magnetic_unit = np.sqrt(4.0 * np.pi * density_unit) * self.velocity_unit
         self.magnetic_unit.convert_to_units("gauss")
+
+    default_units = {
+        "length_unit": "cm",
+        "time_unit": "s",
+        "density_unit": "g/cm**3",
+        "mass_unit": "g",
+        "velocity_unit": "cm/s",
+        "magnetic_unit": "gauss",
+        "temperature_unit": "K",
+    }
+
+    @classmethod
+    def _validate_units_override_keys(cls, units_override):
+        """Check that keys in units_override are consistent with PLUTO's internal
+        normalisations factors.
+        """
+        # For simplicity, it's only allowed for PLUTO's internal normalisations factor,
+        # other combinations might be added in the future
+        valid_keys = {"length_unit", "velocity_unit", "density_unit"}
+        invalid_keys_found = set(units_override.keys()) - valid_keys
+        if invalid_keys_found:
+            raise ValueError(
+                f"units_override contains invalid keys: {invalid_keys_found}. "
+                f"Only allow {valid_keys} to be passed to units_override"
+            )
+
+        super(cls, cls)._validate_units_override_keys(units_override)
```

Comment:
Ah so if I'm not mistaken, you don't actually need to override this method, as you would get the same result by just changing `IdefixVtkDataset.default_units` to only have the three keys you consider valid here `("length_unit", "velocity_unit", "density_unit")`. However, this is probably not a valid approach because then we wouldn't have a default for magnetic and temperature units.

Most importantly, your current implementation is problematic because it prevents using otherwise valid unit normalisation (in `mass_unit` should still be usable).

I believe I designed (or refactored, rather) this class method in yt a while ago. I believe my intention was to allow _extending_ the list of valid units overrides, but it's probably not a good idea to restrict it, compared to the base class.

Reply:
Your point is right. I found the method you designed in the AMRVAC frontend, that's why I commented here that other combinations would be supported in the future. 

I planed to make units_override usable for now. Maybe I don't have to rush to create this PR, so I'm considerring putting this PR on hold until I make this to be a mature feature.

Reply:
My intention is to reduce the dependece on header file `definitions.h` during analysis, and be able to scale the results to other physical system by adapting normalization factor in post-analysis.

Reply:
>  I'm considerring putting this PR on hold until I make this to be a mature feature.

That is fine by me.

> My intention is to reduce the dependece on header file definitions.h during analysis, and be able to scale the results to other physical system by adapting normalization factor in post-analysis.

Sounds like a reasonable approach.